### PR TITLE
feat: URLs

### PR DESCRIPTION
**Pull request by gsgalloway**
Defines classes for parsing, representing, and serializing URLs for the different resource types used by Farcaster. A couple examples:

```typescript
export abstract class BaseURL {
  public abstract toString()
}

export class UserURL extends BaseURL {
  public readonly userId: UserId;
  public static parse(_url: string): Result<UserURL, string> {...}
  public constructor(userId: UserId) {}
  public toString(): string {
    return FarcasterURL.SCHEME + '://' + this.userId.toString();
  }
} 

export class CastURL extends BaseURL {
  public readonly castId: CastId; // includes userId
  public static parse(_url: string): Result<CastURL, string> {...}
  public constructor(castId: CastId) {}
  public toString(): string {
    return FarcasterURL.SCHEME + '://' + this.castId.toString();
  }
}

export function parseUrl(url: string, allowUnrecognized = true): Result<URL, string> {
  ...
}
```

These typed URL instances can be created via either
1. Calling the constructor on a specific BaseURL subclass, passing in the components for that URL as either richly typed objects or raw strings (which will then be parsed). In this way new URLs can be created to address existing resources.
2. Calling the static `parse(url: string)` function on any specific BaseURL subclass, which will attempt to parse the URL string exactly according to that schema, returning an error otherwise. This allows parsing a URL string that is only allowed to be a specific type in the current context (ex: a `follow` message only accepts a UserURL as the follow target)
3. Calling a top-level `parseUrl` function, which walks through all known URL types and returns an instance of the matching BaseURL subclass, if one exists. Otherwise it returns either an error, or a special "unrecognized URL" instance if the URL is well-formed and the caller indicated that valid-but-unrecognized URLs are acceptable. This allows for reacting to arbitrary resources or attaching links to arbitrary resources as message metadata.

The parsing/serialization/validation makes use of [caip-js](https://www.npmjs.com/package/caip) for its implementation of the existing CAIP standards, and caip-js's model is adopted for the implementation of the new `farcaster://` resource types.

caip-js's models aren't especially great to use as developer-facing interfaces, nor do they support URLs that begin with the `chain://` scheme prefix, so this introduces wrappers around the caip-js models that have simpler APIs and recognize our custom URL schemes. For example, `class ChainURL` added in this PR wraps `caip-js`'s `ChainId`. `ChainURL` here represents `'chain://eip155:1'` while its underlying `ChainId` represents just `'eip155:1'`

This is not yet an exhaustive implementation of URL types, nor does it yet integrate the URL types with the farcaster node.

Implemented
- [x] Parse & serialize:
    - [x] URLs describing specific chains (only the chains themselves -- activity on a chain not yet implemented)
    - [x] User URLs
    - [x] Cast URLs
    - [x] URLs with unrecognized 'scheme'
- [x] Tabulated tests for the above

Remaining
- [ ] Integrate these URLs into the rest of the node's source
- [x] HTTP(S) URL support
- [ ] IPFS URL support
- [ ] More chain:// URL support
  - [ ] Resources on unknown chain types (e.g. a Solana address)
  - [x] Eth addresses
  - [ ] CAIP 19 Ethereum assets
  - [ ] Ethereum transactions
  - [ ] Unknown (non-CAIP 19) asset types on Ethereum chains
- [ ] Unrecognized farcaster:// message types
- [x] Parsing/serializing URLs for remaining Farcaster message types
  - [x] Reaction
  - [x] Verified Address
  - [x] Follow
- [x] Integrate and test the URL types with the node's message serializer and deserializer

### CODE REVIEW

**Comment by goksu** [comment]
Overall this makes sense to me and seems straight-forward to implement on clients.

With that being said — one question — if we are going to prefix URLs with user identifiers, it feels like it should be the username and not the user id. I understand that every user can bring their own username from different services or even have none. However URLs are a big part of UX and sticking a random id to it makes it significantly less usable, imo.

cc — @varunsrin

**Comment by varunsrin** [comment]
An application that sees the URI `farcaster://id:123/cast:456` has a deterministic way to find the content. It can look up the AccountRegistry, find the Hub URL for account number 123 and ask for cast 456. 

If we replace id with username, we run into the issue where the current owner of the name isn't the one who created the cast. Applications would have to do  an O(N) lookup where N is the number of times the username has been connected to a different account and find and check each Hub. This N will continue to grow as the network ages and usernames change hands. 

I think its important for the canonical URL's to be as deterministic as possible, and if we need nicer looking URLs for UX applications can have their own URL schemes and a way to translate to the canonical system.

**Comment by varunsrin** [comment]
just took a pass over this again, and it looks good. we should integrate it into some part of the hub codebase so that it is "working" and any breakages will be caught as we keep developing. 

@pfletcherhill defer to you on where best to start integrating this. 

@gsgalloway looks like we might need another rebase to resolve conflicts before we merge.

**Review by pfletcherhill — COMMENTED** [review]
I had to go back and forth between the caip project and this PR to understand what's happening here, but I think I get it now. I'm a little wary of how much code there is to define relatively simple concepts. Let's try integrating it with verifications (referencing an Ethereum address) and recasts (referencing another cast) and see how straightforward it is. Would you like me to open issues for those two integrations, or do you want to try adding those in this PR?

File: src/urls/identifiers.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+  public static parse(id: string): Result<UserIdParams, string> {
+    if (!isValidId(id, this.spec)) {
+      return err(`Invalid ${this.spec.name} provided: ${id}`);
+    }
+    return ok(new UserId(getParams<UserIdParams>(id, this.spec)).toJSON());
+  }
+
+  public static format(params: UserIdParams): string {
+    return joinParams(params as any, this.spec);
```
[the remark this answers is not in the document]
Comment (gsgalloway) [reply src/urls/identifiers.ts:23]: Lifted from caip-js, which itself didn't manage to get the types to work in this case
https://github.com/ChainAgnostic/caip-js/blob/master/src/account.ts#L35

File: src/urls/utils.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
+/**
+ * File pulled from https://github.com/ChainAgnostic/caip-js/blob/b2af86364fc19d2f11deea50858c73567969805a/src/utils.ts
```
[the remark this answers is not in the document]
Comment (gsgalloway) [reply src/urls/utils.ts:2]: Not exported by `caip-js`.

Copied ^ explanation inline

File: src/urls/identifiers.ts [hunk 3]
```diff
@@ -0,0 +1,167 @@
+import { joinParams, getParams, isValidId } from '~/urls/utils';
+import { UserIdSpec, CastIdSpec, CastHashSpec } from '~/urls/specs';
+import { err, ok, Result } from 'neverthrow';
+
+export interface UserIdParams {
```
[the remark this answers is not in the document]
Comment (gsgalloway) [reply src/urls/identifiers.ts]: Happy to move it if you like, but are you sure you want these very in-the-weeds URL-specific types in the shared `types` folder?

I understand aggregating farcaster message types in one file. But I'm curious what the reasoning is for aggregating all typescript types of any kind together into one file

File: src/urls/identifiers.ts [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  namespace: 'id';
```
[the remark this answers is not in the document]
Comment (gsgalloway) [reply src/urls/identifiers.ts]: Updated!

File: src/urls/specs.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
+import type { IdentifierSpec } from 'caip/dist/types';
```
[the remark this answers is not in the document]
Comment (gsgalloway) [reply src/urls/specs.ts]: Commented in module-level docstring on `urls/index.ts`

File: src/urls/identifiers.ts [hunk 6]
```diff
@@ -0,0 +1,167 @@
+import { joinParams, getParams, isValidId } from '~/urls/utils';
```
[the remark this answers is not in the document]
Comment (gsgalloway) [reply src/urls/identifiers.ts]: Moved to 1:1 relationship between new files and the URL types

File: src/urls/identifiers.ts [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+import { err, ok, Result } from 'neverthrow';
+
+export interface UserIdParams {
+  namespace: 'id';
+  value: string;
+}
+
+export type UserIdConstructorArgs = Omit<UserIdParams, 'namespace'>;
+
+export class UserId {
```
[the remark this answers is not in the document]
Comment (gsgalloway) [reply src/urls/identifiers.ts]: Changed to `FarcasterId`

File: src/urls/index.ts [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+export abstract class FarcasterURL extends URL {
+  public static readonly SCHEME = 'farcaster';
+  public readonly scheme = FarcasterURL.SCHEME;
+}
+
+export class UserURL extends FarcasterURL {
+  public readonly userId: UserId;
+
+  public static parse(_url: string): Result<UserURL, string> {
```
[the remark this answers is not in the document]
Comment (gsgalloway) [reply src/urls/index.ts]: You're right! No good reason, not sure why that was there. I think a vestige from when this class had its own `url` member variable and the underscore was to differentiate

File: src/urls/index.ts [hunk 9]
```diff
@@ -0,0 +1,153 @@
+import { Result, err, ok } from 'neverthrow';
+import { CastId, UserId } from '~/urls/identifiers';
+import { parse as rawUriParse, URIComponents } from 'uri-js';
+import { ChainId } from 'caip';
+
+export function parseUrl(url: string, allowUnrecognized = true): Result<URL, string> {
```
[the remark this answers is not in the document]
Comment (gsgalloway) [reply src/urls/index.ts]: Added an eslint rule for enforcing use of arrow functions

**Comment by gsgalloway** [comment]
Working on the integration into `engine`'s verification functions

File: src/engine/engine.verification.test.ts [hunk 10]
```diff
@@ -92,6 +93,51 @@ describe('mergeVerification', () => {
...
   });
 
+  describe('ethereum addresses', () => {
+    test('works when address is checksummed', async () => {
+      // leverage the fact that the happy path test above covers checksummed ethereum addresses
+      const accountURL = ChainAccountURL.parse(genericVerificationAdd.data.body.externalUri)._unsafeUnwrap();
+      expect(accountURL.address).toEqual(ethers.utils.getAddress(accountURL.address));
+    });
+
+    test('works when address is not checksummed', async () => {
```
Comment (gsgalloway) [inline src/engine/engine.verification.test.ts:103]: Important question here - is it important whether or not the node accepts non-canonicalized Account IDs? In this case, either the node must take a stance on checksummed vs non-checksummed ethereum addresses, or it must allow for both to be used and the node must be capable of deduping them internally.

**Comment by gsgalloway** [comment]
> Let's try integrating it with verifications (referencing an Ethereum address) and recasts (referencing another cast) and see how straightforward it is.

This is done

File: src/engine/engine.verification.test.ts (hunk 10 above)
Comment (pfletcherhill) [reply src/engine/engine.verification.test.ts:103]: We should accept all formats and dedup them internally. That's what we do in the signer set at least (lowercase all addresses before comparison or storage).

**Pull request merged**